packages/flutter/lib/src/material/bottom_app_bar.dart

```diff
@@ -174,6 +189,8 @@ class _BottomAppBarState extends State<BottomAppBar> {
   Widget build(BuildContext context) {
     final ThemeData theme = Theme.of(context);
     final bool isMaterial3 = theme.useMaterial3;
+    assert(!isMaterial3 || widget.shape == null); // `shape` not accepted in Material 3
```

Comment:
The information in the comment should go in an assertion message instead, so the developer sees it when they hit the assertion. For example:

```suggestion
    assert(!isMaterial3 || widget.shape == null,
      'The [BottomAppBar.shape] field is not permitted in Material 3.');
```

(Could also throw a `FlutterError` for a fancier error message, but this might be enough.)

Reply:
If throwing a `FlutterError`, would we want to do that in production too? AIUI [`assert`s only run in debug mode](https://docs.flutter.dev/testing/debugging#debug-mode-assertions).

Reply:
No, the check and the throw would both go inside an `assert`. For example, here's [State.context]:
```dart
  BuildContext get context {
    assert(() {
      if (_element == null) {
        throw FlutterError(
          'This widget has been unmounted, so the State no longer has a context (and should be considered defunct). \n'
          'Consider canceling any active work during "dispose" or using the "mounted" getter to determine if the State is still active.',
        );
      }
      return true;
    }());
    return _element!;
  }
```